### PR TITLE
[4][Security] Correctly allow use of IP headers behind Load Balancers, and Not

### PR DESCRIPTION
**Pull request by PhilETaylor**
Joomla 4 version of https://github.com/joomla/joomla-cms/pull/32866 for easy merging.

### This also includes protection of the `/api` endpoint which is specific to Joomla 4

// @joomla/security @zero-24

### CODE REVIEW

**Comment by zero-24** [comment]
I think that should be merged up once merged into 3.x but will let that up to @wilsonge and @rdeutz

**Comment by PhilETaylor** [comment]
This one is slightly different as it has the protection on the new `/api` endpoint as well. 

Anything I can do to make it easier for @wilsonge to "merge up" I will do, easier for him just to hit a merge button :)

File: administrator/components/com_config/forms/application.xml [hunk 1]
```diff
@@ -442,6 +442,19 @@
...
 		label="CONFIG_PROXY_SETTINGS_LABEL">
 
+		<field
+			name="behind_loadbalancer"
+			type="radio"
+			label="COM_CONFIG_FIELD_LOADBALANCER_ENABLE_LABEL"
+			description="COM_CONFIG_FIELD_LOADBALANCER_ENABLE_DESC"
+			layout="joomla.form.field.radio.switcher"
+			default="0"
+		    filter="integer"
```
Comment (brianteeman) [inline administrator/components/com_config/forms/application.xml]:
```suggestion
		        filter="integer"
```

tab

File: administrator/language/en-GB/com_config.ini [hunk 2]
```diff
@@ -132,11 +132,13 @@ COM_CONFIG_FIELD_METADESC_LABEL="Site Meta Description"
 COM_CONFIG_FIELD_METAVERSION_LABEL="Joomla Version"
 COM_CONFIG_FIELD_OFFLINE_IMAGE_LABEL="Offline Image"
 COM_CONFIG_FIELD_OFFLINE_MESSAGE_LABEL="Custom Message"
-COM_CONFIG_FIELD_PROXY_ENABLE_LABEL="Enable Proxy"
-COM_CONFIG_FIELD_PROXY_HOST_LABEL="Proxy Host"
-COM_CONFIG_FIELD_PROXY_PASSWORD_LABEL="Proxy Password"
-COM_CONFIG_FIELD_PROXY_PORT_LABEL="Proxy Port"
-COM_CONFIG_FIELD_PROXY_USERNAME_LABEL="Proxy Username"
+COM_CONFIG_FIELD_LOADBALANCER_ENABLE_DESC="If your site is behind a load balancer or reverse proxy (CloudFlare, Sucuri etc.) enable this setting so that IP addresses and other configurations within Joomla automatically take this into account."
+COM_CONFIG_FIELD_LOADBALANCER_ENABLE_LABEL="Behind Load balancer"
```
Comment (brianteeman) [inline administrator/language/en-GB/com_config.ini]:
```suggestion
COM_CONFIG_FIELD_LOADBALANCER_ENABLE_LABEL="Behind Load Balancer"
```

**Comment by PhilETaylor** [comment]
Code blocks rearranged to satisfy [comments on the Joomla 3 version of this PR](https://github.com/joomla/joomla-cms/pull/32866#issuecomment-808912094).

**Comment by richard67** [comment]
I have tested this item :white_check_mark: successfully on b5a00e5ced9e4e343f47cce1d200d55ae487581f<br /><br />The PR fixes the issue by not allowing IP override when the new option "Behind Load Balancer" is switched off (default).

When the option is switched on, IP override is still allowed (and has to be, otherwise the load balancer or similar would not work).

Tested with help of Firefox' developer tools.

I think we should use this PR as it is and consider to make enhancements like mentioned here https://github.com/joomla/joomla-cms/pull/32866#issuecomment-808912932 with future PR's. For those I am open for "softening" the new feature policy and consider them for 4.0, as they can be considered to be security enhancements, as long as fully b/c. But that's just my personal opinion, I don't speak for anyone else.

Questions: Should there be a hint in release notes to switch the new option on with a clear description of when it has to be done and when not?<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/32869">issues.joomla.org/tracker/joomla-cms/32869</a>.</sub>

**Comment by richard67** [comment]
A postinstall message about the new configuration option will be added by the J3 PR #32866 . This will be merged up into 3.10-dev, too, and so still will be there when people have updated their 3.10 to 4. So nothing more to do here, this PR is fine as it is, as far as I can see.

**Comment by anibalsanchez** [comment]
Hi,

The rest of the fields don't have a description. 

Why this particular field need this description "If your site is behind a load balancer or reverse proxy (CloudFlare, Sucuri etc.) enable this setting so that IP addresses and other configurations within Joomla automatically take this into account."?

a. We have a description to all fields2
b. We don't have a description unless there is a very good reason. ("Behind Load Balancer" is a border case).<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/32869">issues.joomla.org/tracker/joomla-cms/32869</a>.</sub>

**Comment by PhilETaylor** [comment]
Because this is a complex new setting I guess we just need to expound it. 

"CloudFlare, Sucuri " have been removed from the Joomla 3 version of this PR and probably will be from this one too

**Comment by anibalsanchez** [comment]
It is weird at the UX level to have a single field for a border case with a detailed description (even if it a cool description ;-) ). 

I mean we could easily argue that most of the fields in the Global Configuration require knowledge about what you are doing.

**Comment by richard67** [comment]
I feel somehow like sitting in the middle of you now, listening you, and thinking you are both right somehow.

Yes, it's not consistent to have descriptions there.

But yes, I think this setting is one of those where I personally need a description.

**Comment by PhilETaylor** [comment]
Fact is that maintainers were happy with it having a description for merging into Joomla 3.... 

... but then one would hope this is not so "new a feature" in Joomla 4 that people would not need it spelling out to them?

**Comment by richard67** [comment]
> ... but then one would hope this is not so "new a feature" in Joomla 4 that people would not need it spelling out to them?

That could be right, too. I'm ok with both, with or without description in J4.

**Comment by PhilETaylor** [comment]
Closing this as its already been merged into Joomla 3, and so will "eventually" make its way into Joomla 4

At that time (depending on my mood) I will revisit the `/api` endpoint. The rest is already in Joomla 3.

**Comment by wilsonge** [comment]
This has now made it's way into J4. I believe I have also sorted the api endpoint whilst i was doing the conflicts

**Comment by PhilETaylor** [comment]
Thanks - looks to be there correctly and I have tested it and it works as expected both behind a proxy and without a proxy.

**Pull request closed without merging**